services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_aggregate_view_query.rb

```diff
@@ -0,0 +1,207 @@
+        SELECT *
+          FROM aggregate_rows
+      SQL
+    end
+
+    def select_clause
+      <<-SQL
+        SELECT
+          performance.activity_id AS diagnostic_id,
+          activities.name AS diagnostic_name,
+          #{aggregate_by_clause} AS aggregate_id,
+          #{aggregate_sort_clause} AS name,
+          '#{additional_aggregation}' AS group_by,
+          #{specific_select_clause}
+      SQL
+    end
+
+    def from_and_join_clauses
+      <<-SQL
+        FROM lms.pre_post_diagnostic_skill_group_performance_view AS performance
+        JOIN lms.active_classroom_stubs_view AS classrooms ON performance.classroom_id = classrooms.id
+        JOIN lms.classrooms_teachers ON classrooms.id = classrooms_teachers.classroom_id AND classrooms_teachers.role = 'owner'
+        JOIN lms.schools_users ON classrooms_teachers.user_id = schools_users.user_id
+        JOIN lms.activities ON performance.activity_id = activities.id
+        JOIN lms.active_user_names_view AS users ON classrooms_teachers.user_id = users.id
+      SQL
+    end
+
+    def where_clause
+      <<-SQL
+        WHERE
+          #{timeframe_where_clause}
+          #{classroom_ids_where_clause}
+          #{grades_where_clause}
+          #{owner_teachers_only_where_clause}
+          #{relevant_diagnostic_where_clause}
+          #{school_ids_where_clause}
+          #{teacher_ids_where_clause}
+      SQL
+    end
+
+    def rollup_select_columns = "diagnostic_id, diagnostic_name"
+    def specific_select_clause = raise NotImplementedError
+
+    def timeframe_where_clause = "#{relevant_date_column} BETWEEN '#{timeframe_start.to_fs(:db)}' AND '#{timeframe_end.to_fs(:db)}'"
+    def classroom_ids_where_clause = ("AND classrooms.id IN (#{classroom_ids.join(',')})" if classroom_ids.present?)
+    def grades_where_clause = ("AND (classrooms.grade IN (#{grades.map { |g| "'#{g}'" }.join(',')}) #{grades_where_null_clause})" if grades.present?)
+    def grades_where_null_clause = ("OR classrooms.grade IS NULL" if grades.include?('null'))
+    def owner_teachers_only_where_clause = "AND classrooms_teachers.role = '#{ClassroomsTeacher::ROLE_TYPES[:owner]}'"
+    def relevant_diagnostic_where_clause = "AND performance.activity_id IN (#{DIAGNOSTIC_ORDER_BY_ID.join(',')})"
+    def school_ids_where_clause = "AND schools_users.school_id IN (#{school_ids.join(',')})"
+    def teacher_ids_where_clause = ("AND schools_users.user_id IN (#{teacher_ids.join(',')})" if teacher_ids.present?)
+
+    def group_by_clause = "GROUP BY performance.activity_id, activities.name, aggregate_id, #{aggregate_sort_clause}"
+
+    def active_classroom_stubs_view = materialized_view('active_classroom_stubs_view')
+    def active_user_names_view = materialized_view('active_user_names_view')
+    def performance_view = materialized_view('pre_post_diagnostic_skill_group_performance_view')
+
+    def materialized_views = [active_classroom_stubs_view, active_user_names_view, performance_view]
```

Comment:
Small request: Can you move this method to the top of the file below initialize (or below run). Personal preference, but it's a method the subclass is required to implement that powers a lot of the logic. I think putting it up top makes it a little easier to parse.